design/backup-resource-list.md

```diff
@@ -0,0 +1,96 @@
+
+## Goals
+
+- Include a list of backed up resources as metadata in the bucket
+- Enable users to get a view of what resources are included in a backup using the Velero CLI
+
+## Non Goals
+
+- Expose the full manifests of the backed up resources
+
+## Background
+
+As reported in [#396](https://github.com/heptio/velero/issues/396), the information reported in a `velero backup describe <name> --details` command is fairly limited, and does not easily describe what resources a backup contains.
+In order to see what a backup might contain, a user would have to download the backup tarball and extract it.
+This makes it difficult to keep track of different backups in a cluster.
+
+## High-Level Design
+
+After performing a backup, a new file will be created that contains the list of the resources that have been included in the backup.
+This file will be persisted in object storage alongside the backup contents and existing metadata.
+
+A section will be added to the output of `velero backup describe <name> --details` command to view this metadata.
+
+## Detailed Design
+
+### Metadata file
+
+This metadata will be in JSON (or YAML) format so that it can be easily inspected from the bucket outside of Velero tooling, and will contain the API resource and group, namespaces and names of the resources:
+
+```
+apps/v1/Deployment:
+- default/database
+- default/wordpress
+v1/Service:
+- default/database
+- default/wordpress
+v1/Secret:
+- default/database-root-password
+- default/database-user-password
+v1/ConfigMap:
+- default/database
+v1/PersistentVolume:
+- my-pv
+```
+
+The filename for this metadata will be `<backup name>-resource-list.json`.
+The top-level key is the string form of the `schema.GroupResource` type that we currently keep track of in the backup controller code path.
+
+### Changes in Backup controller
+
+The Backupper currently initialises a map to track the `backedUpItems` (https://github.com/heptio/velero/blob/1594bdc8d0132f548e18ffcc1db8c4cd2b042726/pkg/backup/backup.go#L269), this is passed down through GroupBackupper, ResourceBackupper and ItemBackupper where ItemBackupper records each backed up item.
+This property will be moved to the [Backup request struct](https://github.com/heptio/velero/blob/16910a6215cbd8f0bde385dba9879629ebcbcc28/pkg/backup/request.go#L11), allowing the BackupController to access it after a successful backup.
+
+`backedUpItems` currently uses the `schema.GroupResource` as a key for the resource.
+In order to record the API group, version and kind for the resource, this key will be constructed from the object's `schema.GroupVersionKind` in the format `{group}/{version}/{kind}` (e.g. `apps/v1/Deployment`).
+
+The `backedUpItems` map is currently a flat structure and will need to be converted to the nested structure above, grouped by `schema.GroupVersionKind` string.
```

Comment:
You may want to consider decoupling the way this data is stored internally in `pkg/backup` from how it gets written to file for display purposes. Having a flat map makes lookup very efficient during backup, whereas having something like a `map[schema.GroupVersionKind][]string` (which I think is what you're proposing), while closer to what you want for view purposes, would result in less efficient lookups during backup. Not a big deal for small clusters, but it could add some overhead in larger ones. I'm fine sorting this out as you move to implementation.

Reply:
You're right, I think I was envisioning that we would do this conversion when creating the metadata file (i.e. in persistBackup), so I think internally in pkg/backup it would keep the flat structure - I'll update this to make that more clear.

Reply:
Ah, I see - SGTM

Reply:
@skriss reworded a bit to make this more clear